File: examples/go/case06_test.go
package examplesgo

import "testing"

func TestCase06MinimumRevisionRead(t *testing.T) {
	projection := NewProductProjection()
	projection.Put(ProductSnapshot{"Product42", 43, 100})
	if got := ReadCurrentProjection(projection, "Product42"); got.Status != "OK" || got.Snapshot.Revision != 43 {
		t.Fatalf("ordinary read should be able to return revision 43: %+v", got)
	}
	if got := SleepBeforeRead(projection, "Product42", 0); got.Snapshot.Revision != 43 {
		t.Fatal("fixed sleep did not establish revision 44")
	}
	if got := ReadAtLeastRevision(projection, "Product42", 44); got.Status != "NOT_FRESH_ENOUGH" {
		t.Fatalf("43 cannot satisfy minimum 44: %+v", got)
	}
	if got := ReadAtLeastRevision(projection, "Product42", 0); got.Status != "OK" || got.Snapshot.Revision != 43 {
		t.Fatal("eventual read with no minimum may return revision 43")
	}
	projection.Put(ProductSnapshot{"Product42", 44, 120})
	if got := ReadAtLeastRevision(projection, "Product42", 44); got.Status != "OK" || got.Snapshot.Revision != 44 {
		t.Fatal("revision 44 must satisfy minimum 44")
	}
	projection.Put(ProductSnapshot{"Product42", 45, 125})
	if got := ReadAtLeastRevision(projection, "Product42", 44); got.Status != "OK" || got.Snapshot.Price != 125 {
		t.Fatal("revision 45 must satisfy minimum 44 without exact-value polling")
	}
	cache := NewProductProjection()
	cache.Put(ProductSnapshot{"Product42", 43, 100})
	if got := ReadAtLeastRevision(cache, "Product42", 44); got.Status != "NOT_FRESH_ENOUGH" {
		t.Fatal("a stale cache must enforce the same minimum")
	}
}

File: examples/java/Case06Test.java
final class Case06Test {
    public static void main(String[] args) throws Exception {
        Case06.Projection projection = new Case06.Projection();
        projection.put(new Case06.Snapshot("Product42", 43, 100));
        if (Case06.readCurrentProjection(projection, "Product42").snapshot.revision != 43)
            throw new AssertionError("ordinary read may return revision 43");
        if (Case06.sleepBeforeRead(projection, "Product42", 0).snapshot.revision != 43)
            throw new AssertionError("fixed sleep did not prove revision 44");
        if (!Case06.readAtLeastRevision(projection, "Product42", 44).status.equals("NOT_FRESH_ENOUGH"))
            throw new AssertionError("43 cannot satisfy minimum 44");
        if (Case06.readAtLeastRevision(projection, "Product42", 0).snapshot.revision != 43)
            throw new AssertionError("eventual read may return 43");
        projection.put(new Case06.Snapshot("Product42", 44, 120));
        if (Case06.readAtLeastRevision(projection, "Product42", 44).snapshot.revision != 44)
            throw new AssertionError("44 should satisfy minimum 44");
        projection.put(new Case06.Snapshot("Product42", 45, 125));
        if (Case06.readAtLeastRevision(projection, "Product42", 44).snapshot.price != 125)
            throw new AssertionError("45 should satisfy minimum 44");
        Case06.Projection cache = new Case06.Projection();
        cache.put(new Case06.Snapshot("Product42", 43, 100));
        if (!Case06.readAtLeastRevision(cache, "Product42", 44).status.equals("NOT_FRESH_ENOUGH"))
            throw new AssertionError("cache must enforce minimum 44 too");
    }
}
